### PR TITLE
Fix reading new created Hive unpartitioned table within txn

### PR DESCRIPTION
**Pull request by wenleix**
Resolves https://github.com/prestodb/presto/issues/12311

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 1]
```diff
@@ -2005,6 +2005,27 @@ public PartitionStatistics getStatisticsUpdate()
             return statisticsUpdate;
         }
 
+        public Table getAugmentTableForInTransactionRead()
```
Comment (rschlussel) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: nit: either call this getTableForInTransactionRead() or augmentTableForInTransactionRead()

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 2]
```diff
@@ -2005,6 +2005,27 @@ public PartitionStatistics getStatisticsUpdate()
...
+        {
+            if (!this.currentLocation.isPresent()) {
+                // partitioned table
+                return this.table;
+            }
+
+            // For unpartitioned table, this method augments the location field of the table
+            // to the staging location.
+            // This way, if the table is accessed in an ongoing transaction, staged data
+            // can be found and accessed.
```
Comment (rschlussel) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Why is this specific to unpartitioned tables?

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 3]
```diff
@@ -2005,6 +2005,27 @@ public PartitionStatistics getStatisticsUpdate()
...
+            Table table = this.table;
+            String currentLocation = this.currentLocation.get().toString();
+            if (!currentLocation.equals(table.getStorage().getLocation())) {
+                table = Table.builder(table)
+                        .withStorage(storage -> storage.setLocation(currentLocation))
```
Comment (rschlussel) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Is this only used for reads?  Would this have any side effects for writing the table?
e.g. I create a table and then insert into it, will the location be wrong?

**Comment by rschlussel** [comment]
Also looks like the test failures are related.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 1 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Similar method for partition is called `getAugmentedPartitionForInTransactionRead`. Let's call it the same way, `getAugmentedTableForInTransactionRead`. I overlooked `Augment/Augmented`.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 2 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Partitioned tables don't have data stored directly in the table directory.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 3 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Currently we don't allow multiple inserts into a same table / partition within the same transaction.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 1 above)
Comment (rschlussel) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: ah yeah that's fine too.

Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Thanks! Name fixed.

**Comment by wenleix** [comment]
Thanks @arhimondr  and @rschlussel . I amend with two commits to for test failures. Would you mind also taking a look? Thanks!

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 4]
```diff
@@ -1116,7 +1116,13 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
             tableStatistics = new PartitionStatistics(createEmptyStatistics(), ImmutableMap.of());
         }
 
-        metastore.createTable(session, table, principalPrivileges, Optional.of(writeInfo.getWritePath()), false, tableStatistics);
+        metastore.createTable(
+                session,
+                table,
+                principalPrivileges,
+                table.getPartitionColumns().isEmpty() ? Optional.of(writeInfo.getWritePath()) : Optional.empty(),
```
Comment (rschlussel) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: what's in the write path if the table is partitioned?  If it's not meaningful maybe this should be fixed there.  If it is, I don't understand why we can ignore it for this particular case.

Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: @rschlussel : So there is write path and target path in `WriteInfo`.  Basically, write path is the path where data initially write to, and target path is the final location of the data after commit.

- In `STAGE_AND_MOVE_TO_TARGET_DIRECTORY` mode: write path is the staging path, target path is the .... target path :) 
- In `DIRECT_TO_TARGET_XXX` mode: write path and the target path are the same.

For in-txn read, we should use the write path as the location.

Comment (rschlussel) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: I see now that TableAndMore already says that currentLocation is for an unpartitioned table only, so this commit is just enforcing that in someplace where that wasn't doing that before. 
 I'm curious what the failure was when it was provided incorrectly.  It looks to me like the location is only used in places that operate on unpartitioned tables.
Also, could you add a check to the TableAndMore constructor that the value is empty if the table is partitioned?

Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: > I see now that TableAndMore already says that currentLocation is for an unpartitioned table only, so this commit is just enforcing that in someplace where that wasn't doing that before.

Correct. 

> Also, could you add a check to the TableAndMore constructor that the value is empty if the table is partitioned?

Will do, thanks!

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 5]
```diff
@@ -2005,6 +2005,27 @@ public PartitionStatistics getStatisticsUpdate()
             return statisticsUpdate;
         }
 
+        public Table getAugmentedTableForInTransactionRead()
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Curious: Do we need to return non-augmented table for non transactional unpartitioned table read? Could this be generalized as the default `getTable`?

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 6]
```diff
@@ -346,7 +346,7 @@ public synchronized void createTable(
             ConnectorSession session,
             Table table,
             PrincipalPrivileges principalPrivileges,
-            Optional<Path> currentPath,
+            Optional<Path> currentPath, // unpartitioned table only
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Should we consider renaming the variable to show it is for unpartitioned table only? Maybe `currentPathForUnpartitionedTable` so that we don't need a comment?

**Comment by wenleix** [comment]
@shixuan-fan 
> Also, I'm not sure if the commit sequence is correct. Should the other two fixes come in first before the actual fix, or, do they have dependencies?

Yes the first two fixes should come first. I will rebase them before merge.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 5 above)
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Per discussion offline, augmentation should only happen when `tableAction` is not empty.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 6 above)
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: It's a difficult call since this variable name also looks too long.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 7]
```diff
@@ -355,7 +355,14 @@ public synchronized void createTable(
         checkNoPartitionAction(table.getDatabaseName(), table.getTableName());
         SchemaTableName schemaTableName = new SchemaTableName(table.getDatabaseName(), table.getTableName());
         Action<TableAndMore> oldTableAction = tableActions.get(schemaTableName);
-        TableAndMore tableAndMore = new TableAndMore(table, Optional.of(principalPrivileges), currentPath, Optional.empty(), ignoreExisting, statistics, statistics);
+        TableAndMore tableAndMore = new TableAndMore(
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: Unrelated change

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 6 above)
Comment (rschlussel) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: What about `currentPathIfUnpartitioned`?

**Comment by rschlussel** [comment]
Still looks good :)

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 6 above)
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java]: @rschlussel : Thanks for the suggestion! This sounds reasonable. Do you mind apply it in an separate patch? (we should also rename other `Optional<Path> currentPath`, including class fields )
